libs/libarchfpga/src/read_fpga_interchange_arch.cpp

```diff
@@ -1968,74 +2034,157 @@ struct ArchReader {
-
-            type.pinloc.resize({1, 1, 4}, std::vector<bool>(type.num_pins, false));
-            for (e_side side : {TOP, RIGHT, BOTTOM, LEFT}) {
-                for (int pin = 0; pin < type.num_pins; pin++) {
-                    type.pinloc[0][0][side][pin] = true;
-                    type.pin_width_offset[pin] = 0;
-                    type.pin_height_offset[pin] = 0;
-                }
-            }
+            fill_sub_tile(type, sub_tile, pins_size, icount, ocount);
 
-            auto ltype = get_type_by_name<t_logical_block_type>(sub_tile.name, ltypes_);
-            vtr::bimap<t_logical_pin, t_physical_pin> directs_map;
+            type.sub_tiles.push_back(sub_tile);
+        }
+    }
 
-            for (int npin = 0; npin < type.num_pins; npin++) {
-                t_physical_pin physical_pin(npin);
-                t_logical_pin logical_pin(npin);
+    void process_constant_block() {
+        std::vector<std::pair<std::string, std::string>> const_cells{arch_->gnd_cell, arch_->vcc_cell};
 
-                directs_map.insert(logical_pin, physical_pin);
-            }
+        // Create constant complex block
+        t_logical_block_type block;
 
-            sub_tile.equivalent_sites.push_back(ltype);
-
-            type.tile_block_pin_directs_map[ltype->index][sub_tile.index] = directs_map;
-
-            // Assign FC specs
-            int iblk_pin = 0;
-            for (const auto& port : sub_tile.ports) {
-                t_fc_specification fc_spec;
-
-                // FIXME: Use always one segment for the time being.
-                //        Can use the right segment for this IOPIN as soon
-                //        as the RR graph reading from the interchange is complete.
-                fc_spec.seg_index = 0;
-
-                //Apply type and defaults
-                if (port.type == IN_PORT) {
-                    fc_spec.fc_type = e_fc_type::IN;
-                    fc_spec.fc_value_type = default_fc_.in_value_type;
-                    fc_spec.fc_value = default_fc_.in_value;
-                } else {
-                    VTR_ASSERT(port.type == OUT_PORT);
-                    fc_spec.fc_type = e_fc_type::OUT;
-                    fc_spec.fc_value_type = default_fc_.out_value_type;
-                    fc_spec.fc_value = default_fc_.out_value;
-                }
+        block.name = vtr::strdup(const_block_.c_str());
+        block.index = ltypes_.size();
 
-                //Add all the pins from this port
-                for (int iport_pin = 0; iport_pin < port.num_pins; ++iport_pin) {
-                    int true_physical_blk_pin = sub_tile.sub_tile_to_tile_pin_indices[iblk_pin++];
-                    fc_spec.pins.push_back(true_physical_blk_pin);
-                }
+        auto pb_type = new t_pb_type;
+        block.pb_type = pb_type;
 
-                type.fc_specs.push_back(fc_spec);
-            }
+        pb_type->name = vtr::strdup(const_block_.c_str());
+        pb_type->num_pb = 1;
```

Comment:
I see some attributes of the constant block is hard coded. Is it possible to provide more comments? For example, what are the two ports (VCC & GND)? And what are the naming convention for both pb_type and ports? I believe these information is useful when building a legalizer, e.g., check_arch(). As we support multiple file format, we definitely need a legalizer which tells us if the arch database is good enough for downstream functions.

Reply:
I have added a comment on top of the function